### PR TITLE
Add support for grid view

### PR DESCRIPTION
**Pull request by gpapin**
Not too sure if I should base this on develop or master so feel free to edit.
This is pretty much done but still need to address 2 issues

- [ ]  Fix the alphabetical alphabetical sorting asc and desc in the cover view
- [x] Figure out what `B,A <-> A B` does and fix it in the cover view
- [ ]  Fix tooltip alignment on large displays
 
Feedback and pointers welcome.

### CODE REVIEW

**Comment by OzzieIsaacs** [comment]
The B,A <-> A B shouldn't be visible. It's made for the authors view (displaying surname, first-name or first name surname)

**Comment by gpapin** [comment]
I think I found a solution, using the  the `isotop` lib that is included should solve all my problems.

**Comment by gpapin** [comment]
Ok the default theme works but in Caliblur the sorting is still broken...

**Comment by gpapin** [comment]
Ok so the 2 remaining issues are linked

1. On the default theme: I use the `.book` class on each items filtering through isotope works fine but the tooltips are a bit off when the display is too wide, presumably because isotope is making each `book` element a bit wider to fit the screen (instead of using a flex-display for example).

2. On Caliblur theme the tooltip seems to work because the theme doesn't use a fixed number of items per rows so the `book` don't expand and the tooltip is where it should be. However there is some other css rules that mess with isotope that breaks the filtering... if I change my `book` css class to `item` for example the filtering start working again! But of course that breaks the tooltips as we see the same issue that we get from the default theme...

**Comment by OzzieIsaacs** [comment]
What about the crazy idea of having the badge in the upper left corner:
grafik

**Comment by OzzieIsaacs** [comment]
The sorting problem arises from these lines in caliblur.css.min:
```
.book {
    width: 225px;
    max-width: 225px;
    //position: relative !important;
    //left: auto !important;
    //top: auto !important;
    //-webkit-transform: none !important;
    //    -ms-transform: none !important;
    //        transform: none !important;
    min-width: 225px;
    display: block;
}
```
If you comment out the lines like in the code snippet the searching works

**Comment by gpapin** [comment]
Thanks for looking in to this. I'll try out your fix and see if that affects anything else. I would be curious to know why those rules were there in the first place, but it this fist it I'll probably override it just in the grid view so it won't affect any other places this is used.

Regarding the left alignment, like you said this is crazy XD. 
Seriously though, I think it's against common practices, in the western world (were we read from left to right, top bottom), any badges you'll see over the internet are on the right hand side of what it represent. Changing that might be confusing.
On the other hand, this would be a very small change and maybe one that could be implemented in a custom theme ? I've seen that there are already some discutions about theme and this isn't easy but what I've seen in some other app is rather than changing the whole theme drastically like Caliblur does, some app let you add custom css that runs after everything else. This is where I would put those changes.

**Comment by OzzieIsaacs** [comment]
The thing I realized about the css, it's shortening the titles of the book, without the css the book titles can be very ling and go into the next book.

File: cps/static/js/filter_grid.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+var $list = $('#list').isotope({
+  itemSelector: '.book',
+  layoutMode: 'fitRows',
```
Comment (OzzieIsaacs) [inline cps/static/js/filter_grid.js]: Codacy Issue found: [Expected indentation of 4 spaces but found 2.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846309)

**Comment by OzzieIsaacs** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 19
           

Complexity increasing per file
==============================
- cps/ub.py  1
- cps/static/js/filter_grid.js  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846309&bid=16085445)

File: cps/static/js/filter_grid.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    $list.isotope({
+        sortBy: 'name',
+        sortAscending: false
+    });
+    return;
+});
+
+$("#asc").click(function() {
+    $list.isotope({
+        sortBy: 'name',
```
Comment (OzzieIsaacs) [inline cps/static/js/filter_grid.js]: Codacy Issue found: [Strings must use doublequote.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846309)

File: cps/static/js/filter_grid.js (hunk 1 above)
Comment (OzzieIsaacs) [inline cps/static/js/filter_grid.js]: Codacy Issue found: [Strings must use doublequote.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846309)

File: cps/static/js/filter_grid.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+var $list = $('#list').isotope({
```
Comment (OzzieIsaacs) [inline cps/static/js/filter_grid.js]: Codacy Issue found: [Strings must use doublequote.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846309)

File: cps/static/js/filter_grid.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  itemSelector: '.book',
+  layoutMode: 'fitRows',
+  getSortData: {
+    title: '.title',
```
Comment (OzzieIsaacs) [inline cps/static/js/filter_grid.js]: Codacy Issue found: [Strings must use doublequote.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846309)

File: cps/static/js/main.js [hunk 5]
```diff
@@ -239,4 +239,17 @@ $(function() {
...
+    $(".update-view").click(function(e) {
+        var target = $(this).data('target');
+        var view = $(this).data('view');
+
+        e.preventDefault();
+        e.stopPropagation();
+        data = {};
+        data[target] = view;
+        console.debug("Updating view data: ", data);
+        $.post( "/ajax/view", data).done(function( result ) {
```
Comment (OzzieIsaacs) [inline cps/static/js/main.js]: Codacy Issue found: ['data' is not defined.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846309)

File: cps/static/js/filter_grid.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  }
```
Comment (OzzieIsaacs) [inline cps/static/js/filter_grid.js]: Codacy Issue found: [Expected indentation of 4 spaces but found 2.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846309)

File: cps/static/js/main.js [hunk 7]
```diff
@@ -239,4 +239,17 @@ $(function() {
         $(".discover .row").isotope("layout");
     });
 
+    $(".update-view").click(function(e) {
+        var target = $(this).data('target');
+        var view = $(this).data('view');
```
Comment (OzzieIsaacs) [inline cps/static/js/main.js]: Codacy Issue found: [Strings must use doublequote.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846309)

File: cps/static/js/filter_grid.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+var $list = $('#list').isotope({
+  itemSelector: '.book',
```
Comment (OzzieIsaacs) [inline cps/static/js/filter_grid.js]: Codacy Issue found: [Strings must use doublequote.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846309)

File: cps/static/js/main.js [hunk 9]
```diff
@@ -239,4 +239,17 @@ $(function() {
         $(".discover .row").isotope("layout");
     });
 
+    $(".update-view").click(function(e) {
+        var target = $(this).data('target');
```
Comment (OzzieIsaacs) [inline cps/static/js/main.js]: Codacy Issue found: [Strings must use doublequote.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846309)

File: cps/static/js/filter_grid.js (hunk 8 above)
Comment (OzzieIsaacs) [inline cps/static/js/filter_grid.js]: Codacy Issue found: [Expected indentation of 4 spaces but found 2.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846309)

File: cps/static/js/filter_grid.js [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+  layoutMode: 'fitRows',
+  getSortData: {
```
Comment (OzzieIsaacs) [inline cps/static/js/filter_grid.js]: Codacy Issue found: [Expected indentation of 4 spaces but found 2.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846309)

File: cps/static/js/main.js [hunk 11]
```diff
@@ -239,4 +239,17 @@ $(function() {
...
+        var view = $(this).data('view');
+
+        e.preventDefault();
+        e.stopPropagation();
+        data = {};
```
Comment (OzzieIsaacs) [inline cps/static/js/main.js]: Codacy Issue found: ['data' is not defined.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846309)

File: cps/static/js/filter_grid.js [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    var character = this.innerText;
+    $(".sortable").each(function() {
+        if (this.attributes["data-id"].value.charAt(0).toUpperCase() !== character) {
+            $(this).hide();
+        } else {
+            $(this).show();
+        }
+    });
+    // We need to trigger the resize event to have all the grid item to re-align.
+    window.dispatchEvent(new Event('resize'));
```
Comment (OzzieIsaacs) [inline cps/static/js/filter_grid.js]: Codacy Issue found: [Strings must use doublequote.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846309)

File: cps/static/js/main.js (hunk 5 above)
Comment (OzzieIsaacs) [inline cps/static/js/main.js]: Codacy Issue found: ['result' is defined but never used.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846309)

File: cps/static/js/main.js [hunk 13]
```diff
@@ -239,4 +239,17 @@ $(function() {
...
+        data[target] = view;
+        console.debug("Updating view data: ", data);
```
Comment (OzzieIsaacs) [inline cps/static/js/main.js]: Codacy Issue found: [Unexpected console statement.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846309)

File: cps/static/js/main.js [hunk 14]
```diff
@@ -239,4 +239,17 @@ $(function() {
...
     });
 
+    $(".update-view").click(function(e) {
+        var target = $(this).data('target');
+        var view = $(this).data('view');
+
+        e.preventDefault();
+        e.stopPropagation();
+        data = {};
+        data[target] = view;
```
Comment (OzzieIsaacs) [inline cps/static/js/main.js]: Codacy Issue found: ['data' is not defined.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846309)

File: cps/static/js/filter_grid.js (hunk 4 above)
Comment (OzzieIsaacs) [inline cps/static/js/filter_grid.js]: Codacy Issue found: [Expected indentation of 8 spaces but found 4.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846309)

File: cps/static/js/main.js (hunk 13 above)
Comment (OzzieIsaacs) [inline cps/static/js/main.js]: Codacy Issue found: ['data' is not defined.](https://app.codacy.com/manual/Ozzie-IsaacsOrganization/calibre-web/pullRequest?prid=4846309)

**Comment by gpapin** [comment]
I like that! I'm going to update at some point.
Is there a way to run it locally so I can test I haven't missed anything before pushing ?

**Comment by OzzieIsaacs** [comment]
Partly: Codacy is using different "linter" to do the checking. 
For javascript it's ESLint and there is a config file in the root folder. For python code I don't have a config. 
I'm sorry about all the emails, this was better in the past

**Comment by gpapin** [comment]
Yeah don't worry gmail keeps it tidy in a thread anyway :)
I'm under the water at the moment but I haven't forgotten this feature by the way.

It would be nice to have a `make test` or `make lint` that would run the lint in a docker container. This is what I got for my other projects and it prevent me from having to go back and forth

**Comment by gpapin** [comment]
I guess this can be close now that https://github.com/janeczku/calibre-web/pull/1306 has been merged

**Pull request closed without merging**